server/src-lib/Hasura/GraphQL/Resolve/Types.hs

```diff
@@ -161,14 +160,64 @@ partialSQLExpToUnresolvedVal = \case
   PSESessVar ty sessVar -> UVSessVar ty sessVar
   PSESQLExp s           -> UVSQL s
 
--- A value that will be converted to an sql expression eventually
+-- | A value that will be converted to an sql expression eventually
 data UnresolvedVal
-  -- From a session variable
   = UVSessVar !(PGType PGScalarType) !SessVar
-  -- This is postgres
+  -- | a SQL value literal that can be parameterized over
   | UVPG !AnnPGVal
-  -- This is a full resolved sql expression
+  -- | an arbitrary SQL expression, which /cannot/ be parameterized over
   | UVSQL !S.SQLExp
   deriving (Show, Eq)
 
 type AnnBoolExpUnresolved = AnnBoolExp UnresolvedVal
+
+-- | Tracks whether or not a query is /reusable/. Reusable queries are nice, since we can cache
+-- their resolved ASTs and avoid re-resolving them if we receive an identical query. However, we
+-- can’t always safely reuse queries if they have variables, since some variable values can affect
+-- the generated SQL. For example, consider the following query:
+--
+-- > query users_where($condition: users_bool_exp!) {
+-- >   users(where: $condition) {
+-- >     id
+-- >   }
+-- > }
+--
+-- Different values for @$condition@ will produce completely different queries, so we can’t reuse
+-- its plan (unless the variable values were also all identical, of course, but we don’t bother
+-- caching those).
+--
+-- If a query does turn out to be reusable, we build up a 'ReusableVariableTypes' value that maps
+-- variable names to their types so that we can use a fast path for validating new sets of
+-- variables (namely 'Hasura.GraphQL.Validate.validateVariablesForReuse').
+data QueryReusability
+  = Reusable !ReusableVariableTypes
+  | NotReusable
+  deriving (Show, Eq)
+
+instance Semigroup QueryReusability where
+  Reusable a <> Reusable b = Reusable (a <> b)
+  _          <> _          = NotReusable
+instance Monoid QueryReusability where
+  mempty = Reusable mempty
+
+class (MonadError QErr m) => MonadResolve m where
```

Comment:
I'm wondering about the use of `MonadError` constraint as both `recordVariableUse` and `markNotReusable` functions don't need this. From what I understand, the `MonadError` constraint is needed for all the `as..ValM` family of functions. Does it make sense to bring the requirement of those functions into this type class? If I'm not wrong, all of these do a `throw500`, so maybe this typeclass can have something like `throwResolutionError`?

Reply:
That’s fair; I debated including this superclass. Ultimately the main reason I decided to include it is because, currently, essentially *everything* in `Resolve/` needs to be able to throw (again, due to the issue described in #2801), and having every function’s constraint say `(MonadError QErr m, MonadResolve m)` just felt redundant. I’m happy to split it back out, though, if you think that’s better; I don’t really have an opinion either way.

Reply:
I think it is okay, doesn't significantly improve the status quo.